From 87249049a9da3b009bc2358c23aea39f38bf32e6 Mon Sep 17 00:00:00 2001
From: saikumaru <saikumaru@ami.com>
Date: Tue, 9 Sep 2025 11:08:43 +0530
Subject: [PATCH] Modify driver pmbus Add conpatible table

Add compatible table to pmbus.c, so we can use multiple source in dts.

Signed-off-by: saikumaru <saikumaru@ami.com>
---
 drivers/hwmon/pmbus/pmbus.c | 48 +++++++++++++++++++++++++++++++++++--
 1 file changed, 46 insertions(+), 2 deletions(-)

diff --git a/drivers/hwmon/pmbus/pmbus.c b/drivers/hwmon/pmbus/pmbus.c
index 59424dc518c8..114986268bf9 100644
--- a/drivers/hwmon/pmbus/pmbus.c
+++ b/drivers/hwmon/pmbus/pmbus.c
@@ -13,6 +13,7 @@
 #include <linux/mutex.h>
 #include <linux/i2c.h>
 #include <linux/pmbus.h>
+#include <linux/of_device.h>
 #include "pmbus.h"
 
 struct pmbus_device_info {
@@ -174,7 +175,12 @@ static int pmbus_probe(struct i2c_client *client)
 	if (!info)
 		return -ENOMEM;
 
-	device_info = (struct pmbus_device_info *)i2c_match_id(pmbus_id, client)->driver_data;
+	if (dev->of_node)
+		device_info = of_device_get_match_data(dev);
+	else
+		device_info = (struct pmbus_device_info *)i2c_match_id(pmbus_id, client)->driver_data;
+
+
 	if (device_info->flags) {
 		pdata = devm_kzalloc(dev, sizeof(struct pmbus_platform_data),
 				     GFP_KERNEL);
@@ -249,11 +255,49 @@ static const struct i2c_device_id pmbus_id[] = {
 
 MODULE_DEVICE_TABLE(i2c, pmbus_id);
 
+static const struct of_device_id __maybe_unused pmbus_of_match[] = {
+	{.compatible = "pmbus", .data = &pmbus_info_zero,},
+	{.compatible = "adp4000", .data = &pmbus_info_one,},
+	{.compatible = "bmr310", .data = &pmbus_info_one_status,},
+	{.compatible = "bmr453", .data = &pmbus_info_one,},
+	{.compatible = "bmr454", .data = &pmbus_info_one,},
+	{.compatible = "bmr456", .data = &pmbus_info_one,},
+	{.compatible = "bmr457", .data = &pmbus_info_one,},
+	{.compatible = "bmr458", .data = &pmbus_info_one_status,},
+	{.compatible = "bmr480", .data = &pmbus_info_one_status,},
+	{.compatible = "bmr490", .data = &pmbus_info_one_status,},
+	{.compatible = "bmr491", .data = &pmbus_info_one_status,},
+	{.compatible = "bmr492", .data = &pmbus_info_one,},
+	{.compatible = "dps460", .data = &pmbus_info_one_skip,},
+	{.compatible = "dps650ab", .data = &pmbus_info_one_skip,},
+	{.compatible = "dps800", .data = &pmbus_info_one_skip,},
+	{.compatible = "max20796", .data = &pmbus_info_one,},
+	{.compatible = "mdt040", .data = &pmbus_info_one,},
+	{.compatible = "ncp4200", .data = &pmbus_info_one,},
+	{.compatible = "ncp4208", .data = &pmbus_info_one,},
+	{.compatible = "pdt003", .data = &pmbus_info_one,},
+	{.compatible = "pdt006", .data = &pmbus_info_one,},
+	{.compatible = "pdt012", .data = &pmbus_info_one,},
+	{.compatible = "pmbus", .data = &pmbus_info_zero,},
+	{.compatible = "sgd009", .data = &pmbus_info_one_skip,},
+	{.compatible = "tps40400", .data = &pmbus_info_one,},
+	{.compatible = "tps544b20", .data = &pmbus_info_one,},
+	{.compatible = "tps544b25", .data = &pmbus_info_one,},
+	{.compatible = "tps544c20", .data = &pmbus_info_one,},
+	{.compatible = "tps544c25", .data = &pmbus_info_one,},
+	{.compatible = "udt020", .data = &pmbus_info_one,},
+	{}
+};
+MODULE_DEVICE_TABLE(of, pmbus_of_match);
+
+
+
 /* This is the driver that will be inserted */
 static struct i2c_driver pmbus_driver = {
 	.driver = {
 		   .name = "pmbus",
-		   },
+		   .of_match_table = pmbus_of_match,
+	},
 	.probe = pmbus_probe,
 	.id_table = pmbus_id,
 };
-- 
2.47.1

